### PR TITLE
Refactor CreateEffectWarhead ImpactTypes

### LINKED ISSUE #11848: Refactor warhead ImpactTypes
We currently have too many of them, some make hardcoded assumptions, some use the soon-to-be-deprecated `IsWater` check, and most ignore `TargetTypes` completely.

### PR DESCRIPTION
**Pull request by reaperrr**
Not only were there quite a few of them, more importantly they also ignored target validity.

This PR solves both.
As side-effect, it fixes the issue in RA where a weapon hitting a cell occupied by a submerged sub would display an explosion instead of a water splash.

TD and D2k didn't use custom impact type settings, so all they need are a few yaml changes to cover the fact that `Warhead.ValidTargets` does not contain `Air` by default.

Closes #11848.

### CODE REVIEW

**Comment by GraionDilach** [comment]
This looks like the AS solution. If a followup then shifts the impacttypes to warheads, then this looks really promising.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 1]
```diff
@@ -828,6 +828,53 @@ internal static void UpgradeWeaponRules(ModData modData, int engineVersion, ref
...
+					if (node.Key.StartsWith("Warhead", StringComparison.Ordinal) && node.Value.Value == "CreateEffect")
+					{
+						var validImpactTypes = node.Value.Nodes.FirstOrDefault(n => n.Key == "ValidImpactTypes");
+						var invalidImpactTypes = node.Value.Nodes.FirstOrDefault(n => n.Key == "InvalidImpactTypes");
+						if (validImpactTypes != null)
+						{
+							Console.WriteLine("CreateEffectWarhead now uses Valid-/InvalidTargets instead of Valid-/InvalidImpactTypes.");
+							Console.WriteLine("Please check whether you need to make manual adjustments.");
+
+							var validTargets = "";
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: It would be cleaner to make this a `List<string>` and then use `FieldSaver.FormatValue` to turn it into a yaml string at the end.

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 2]
```diff
@@ -89,6 +69,10 @@ public bool GetDirectHit(World world, CPos cell, WPos pos, Actor firedBy, bool c
 				if (checkTargetType && !IsValidAgainst(victim, firedBy))
 					continue;
 
+				var healthInfo = victim.Info.TraitInfoOrDefault<HealthInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: This can simplify down to `if (!victim.Info.HasTraitInfo<HealthInfo>())`

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 3]
```diff
@@ -132,10 +116,22 @@ public bool IsValidImpact(WPos pos, Actor firedBy)
 				return false;
 
 			var impactType = GetImpactType(world, targetTile, pos, firedBy);
-			if (!ValidImpactTypes.HasFlag(impactType) || InvalidImpactTypes.HasFlag(impactType))
-				return false;
+			var validImpact = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: This can be simplify down to:
```csharp
var validImpact = false;
switch (impactType)
{
	case ImpactType.TargetHit:
		return true;
	case ImpactType.Air:
		return IsValidTarget(new string[] { "Air" });
	case ImpactType.Ground:
		var tileInfo = world.Map.GetTerrainInfo(targetTile);
		return IsValidTarget(tileInfo.TargetTypes);
	default:
		return false;
}
```

**Review by pchote — COMMENTED** [review]
Here's a couple of code style suggestions to start with.   I haven't tested this ingame, but the yaml changes look like I would expect.

**Comment by pchote** [comment]
Needs a rebase now, too.

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs (hunk 3 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: I assume you mean `switch (validImpact)`?

Comment (pchote) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: Nope, I mean `switch (impactType)` but the `var validImpact` can be deleted.

**Comment by reaperrr** [comment]
Rebased and updated.

**Review by pchote — COMMENTED** [review]
Weapons hitting boats in RA show both splash and explosion animations.  I don't think this is intentional?

**Comment by reaperrr** [comment]
> Weapons hitting boats in RA show both splash and explosion animations. I don't think this is intentional?

Only for DepthCharge vs. submerged subs, not the rest. I'll take a look.

**Comment by reaperrr** [comment]
Turns out that the original version of this had a rather significant limitation:

If `GetDirectHit` only found invalid actors, it would return `false`, which would have the same result as if there weren't any actors on that cell at all: `GetImpactType` would return `ImpactType.Ground`. As a result, even if there were 'blacklisted', invalid actors on that cell, as long as the cell itself contained a valid target type, the warhead would still trigger.
Playing the animation + sound even when a cell contains explicitly invalid actors is counter-intuitive, and broke RA, since `Water` is obviously valid for splashes, so they would play despite `Ship` being blacklisted.

I fixed this by upgrading the `GetDirectHit` boolean to `GetDirectHitTargetType` using a new `ImpactTargetType` enum with the types `ValidActor`, `InvalidActor` and `NoActor`, and returning `ImpactType.None` in `GetImpactType` if `ImpactTargetType` is `InvalidActor`.

Additionally, I forgot to black-list `Structure` for splashes in RA, so it would have played on ship yard/sub pen even when the code was fixed.

I'm pretty confident this now works as intended and covers all edge cases (at least it does in RA, which is our most complex test-case for this), but you never know, so review carefully.

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 4]
```diff
@@ -56,49 +53,53 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
...
 				var activeShapes = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled);
-				if (activeShapes.Any(i => i.Info.Type.DistanceFromEdge(pos, victim).Length <= 0))
-					return true;
+				var directHit = activeShapes.Any(i => i.Info.Type.DistanceFromEdge(pos, victim).Length <= 0);
+
+				// If the warhead landed outside the actor's hit-shape(s), we need to skip the rest so it won't be considered an invalidHit
+				if (!directHit)
+					continue;
+
+				if (directHit && (!checkTargetValidity || (checkTargetValidity && IsValidAgainst(victim, firedBy))))
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: Simplifies to `if (!checkTargetValidity || IsValidAgainst(victim, firedBy))`.

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 5]
```diff
@@ -143,10 +144,18 @@ public bool IsValidImpact(WPos pos, Actor firedBy)
...
-			if (!ValidImpactTypes.HasFlag(impactType) || InvalidImpactTypes.HasFlag(impactType))
-				return false;
-
-			return true;
+			switch (impactType)
+			{
+				case ImpactType.TargetHit:
+					return true;
+				case ImpactType.Air:
+					return IsValidTarget(new string[] { "Air" });
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: Consider caching this array.

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 6]
```diff
@@ -54,51 +51,57 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
 				VictimScanRadius = Util.MinimumRequiredVictimScanRadius(rules);
 		}
 
+		readonly string[] targetTypeAir = new string[] { "Air" };
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: Style nit: `static readonly string[] TargetTypeAir = { "Air" };`

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 7]
```diff
@@ -54,51 +51,57 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
...
-
 			// Matching target actor
-			if (ValidImpactTypes.HasFlag(ImpactType.TargetHit) && GetDirectHit(world, cell, pos, firedBy, true))
-				return ImpactType.TargetHit;
+			if (VictimScanRadius > WDist.Zero)
+			{
+				var targetType = GetDirectHitTargetType(world, cell, pos, firedBy, true);
+				if (targetType == ImpactTargetType.ValidActor)
+					return ImpactType.TargetHit;
+				else if (targetType == ImpactTargetType.InvalidActor)
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: Style nit: you can omit the `else` here (use `if` by itself) because the first branch returns.

File: OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs [hunk 8]
```diff
@@ -54,51 +51,57 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
...
+				if (!checkTargetValidity || IsValidAgainst(victim, firedBy))
+					return ImpactTargetType.ValidActor;
+
+				// If we got here, it must be an invalid target
+				invalidHit = true;
 			}
 
-			return false;
+			// There was at least a single direct hit, but none on valid target(s)
+			if (invalidHit)
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs]: Style nit: `return invalidHit ? ImpactTargetType.InvalidActor : ImpactTargetType.NoActor;`

**Review by pchote — COMMENTED** [review]
A couple of minor style nits for you while I test this ingame.

**Comment by pchote** [comment]
* Planes/helicopters crashing into the ground/water disappear instead of exploding.
* Killed boats show both an explosion and splash.

**Comment by reaperrr** [comment]
Updated.

The boats showing both as well as the missing flak cannon air explosion (noticed by myself) were just yaml issues.

> Planes/helicopters crashing into the ground/water disappear instead of exploding.

This is basically a pecularity of aircraft husks: since they are untargetable, but *do* have `Health` and `HitShape` (which are both needed to make `self.Kill` work in `FallToEarth`), the victim check would never `continue` early (impact pos will obviously always be within their hitshape) and would always return `InvalidActor` on them.
Added
```
				if (!AffectsParent && victim == firedBy)
					continue;
```
to the beginning of the victim check loop, which also has the nice default side-effect of probably improving performance a little if the `firedBy` actor is within the `VictimScanRadius`.

**Review by pchote — APPROVED** [review]

**Comment by jrb0001** [comment]
The hardcoded value "Air" should be mentioned in the documentation for `ValidTargets` and `InvalidTargets`. An even better way would be to make it configurable, but I don't see how to do that properly at right now.

As Graion has already mentioned, this logic should be moved to `Warhead` to reuse it in other warheads. A followup for that might be a good idea to avoid scope creep of this PR.

**Comment by reaperrr** [comment]
> The hardcoded value "Air" should be mentioned in the documentation for ValidTargets and InvalidTargets.

Doing that now would misleading, because it's only true for the CEWH (for now). I'll add a comment about it to `AirThreshold`'s desc, though.

**Comment by GraionDilach** [comment]
I had thought about this during the time I set up the WarheadAS class back then, but my other idea (which I felt messy in a different way) was to exposing the targettypes of air through expanding all theater yamls with an Air terraintype - which would shift the hardcode to a separate level.

**Comment by reaperrr** [comment]
Updated.

> I had thought about this during the time I set up the WarheadAS class back then, but my other idea (which I felt messy in a different way) was to exposing the targettypes of air through expanding all theater yamls with an Air terraintype - which would shift the hardcode to a separate level.

I think compared to that, simply adding an `AirTargetType(s)` field to Warhead would be the lesser evil.

**Comment by jrb0001** [comment]
The description looks good to me.

> I think compared to that, simply adding an `AirTargetType(s)` field to Warhead would be the lesser evil.

Wouldn't it be better to add a boolean like `HitEmptyAir` instead? Adding a new field just to compare it with another field sounds a bit weird to me.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 9]
```diff
@@ -900,6 +900,53 @@ internal static void UpgradeWeaponRules(ModData modData, int engineVersion, ref
...
+								validImpactTypes.Value.Value = validTargets.JoinWith(", ");
+								RenameNodeKey(validImpactTypes, "ValidTargets");
+							}
+							else
+								node.Value.Nodes.Remove(validImpactTypes);
+						}
+						else
+						{
+							// 'Air' is not part of the internal warhead ValidTargets default, but was part of the ValidImpactTypes default.
+							node.Value.Nodes.Add(new MiniYamlNode("ValidTargets", "Ground, Water, Air"));
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: What if a `ValidTargets` yaml node already exists?

**Comment by abcdefg30** [comment]
Depth charges don't damage surfaced subs, I think that should be changed.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs (hunk 9 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Normally it shouldn't, since ValidTargets previously had no effect for CEWH, but I'll add a safety-net just in case.

**Comment by reaperrr** [comment]
> Depth charges don't damage surfaced subs, I think that should be changed.

It's like that on bleed too, though. Are you sure?

**Comment by abcdefg30** [comment]
Yes, as you can dodge depth charges by force firing with the submarine, which looks bogus to me. But if it's already that way on bleed, you can leave it as is, I just thought you touch that anyway in this PR.

**Comment by pchote** [comment]
If that is a regression vs `release-20170527` then we will still need to fix it before the playtest.

**Comment by reaperrr** [comment]
> If that is a regression vs release-20170527 then we will still need to fix it before the playtest.

It isn't, it's the same in last release. 
I agree with @abcdefg30 that it looks bogus, but it would definitely affect balance to fix it now, since DepthCharge is stronger vs. subs than the Gunboat cannon.

**Comment by reaperrr** [comment]
Updated upgrade rule.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/aad16a0588898957ab59b987c5b2946e7b59e587...fbeba2c5a7c09b72d0a773abd67484ee50b16068)
